### PR TITLE
Add View tests

### LINKED ISSUE #43: Add view tests
Nothing fancy, just execute code in the views as a form of "smoke" test to prove all works properly across all Django/Python configurations.

### PR DESCRIPTION
**Pull request by kasulani**
* add tests for RateView
* fixes #43

### CODE REVIEW

File: pinax/ratings/tests/test_views.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    # tests for the RateView
+
+    def setUp(self):
+        self.test_user = self.make_user(username="test_user")
+        self.forester = Car.objects.create(name="Subaru Forester")
+
+    def test_post_rating_with_invalid_content_type_id(self):
+        # test line with code snippet:
+        #   get_object_or_404(ContentType, pk=self.kwargs.get("content_type_id"))
+        #
```
Comment (grahamu) [inline pinax/ratings/tests/test_views.py]: Yes, it's good to have a comment explaining the test goal!

Two comments here: Please use the "docstring" triple double-quote format as described in the Pinax [How to Contribute -
 Coding Style](http://pinaxproject.com/pinax/how_to_contribute/):

```
"""
docstring
"""
```

Secondly, it's not usually good practice to show lines from the tested code. Tests really shouldn't know what is going on in the underlying tested code, but rather the expected result. In this case, the test is proving the view returns 404 if the generic object is not found. There's no need to copy the view code into the docstring. `"""Ensure view returns 404 if object is not found"""` is a reasonably good docstring here.

This commentary applies to the other test methods as well. Their initial comments should be converted to docstrings.

Docstrings are useful because often test runners display the failing test docstring, giving you a better idea about what failed than just the test name.

Docstrings are also helpful to developers trying to understand the package as a whole.

File: pinax/ratings/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                content_type_id=400,
+                object_id=1,
+                data={
+                    "rating": 3,
+                    "category": "handling"
+                },
+            )
+            self.response_404(response)
+
+    def test_post_rating_with_invalid_object_id(self):
```
Comment (grahamu) [inline pinax/ratings/tests/test_views.py]: I understand your thinking writing this test... you're being careful and thorough. However, since `test_post_rating_with_invalid_content_type_id()` proved that bad "generic" data returns a 404, you really don't need to prove it again.

Why? The url kwargs `content_type_id` and `object_id` are always used as a pair, and if one is invalid and we get a 404 that's good enough. You can remove this test as the prior test did the job.

At the same time, rename the first test such that it's clear you're testing for response to an invalid generic object, not the actual url kwargs.

File: pinax/ratings/tests/test_views.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                content_type_id=ContentType.objects.get_for_model(self.forester).pk,
+                object_id=self.forester.pk,
+                data={
+                    "rating": 10,
+                    "category": "handling"
+                },
+            )
+            self.response_403(response)
+
+    def test_post_rating_with_valid_data(self):
```
Comment (grahamu) [inline pinax/ratings/tests/test_views.py]: docstring wanted here. What exactly is being proven?

File: pinax/ratings/tests/test_views.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                },
+            )
+            # assert http response code
+            self.response_200(response)
+            # assert the data returned is as expected
+            self.assertContext("user_rating", 3)
+            self.assertContext("category", "handling")
+            self.assertContext("overall_rating", 3.0)
+            self.assertContext("content_type_id", str(content_type_id))
+            self.assertContext("object_id", str(object_id))
```
Comment (grahamu) [inline pinax/ratings/tests/test_views.py]: These last two lines aren't really needed. The main focus of this test is getting a 200 with correct values for "user_rating", "category", and "overall_rating" in context.

File: pinax/ratings/tests/test_views.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        # tests the condition:
+        #   if rating_input not in range(NUM_OF_RATINGS + 1):
+        #
+        with self.login(self.test_user):
+            response = self.post(
+                "pinax_ratings:rate",
+                content_type_id=ContentType.objects.get_for_model(self.forester).pk,
+                object_id=self.forester.pk,
+                data={
+                    "rating": 10,
```
Comment (grahamu) [inline pinax/ratings/tests/test_views.py]: This works for sure, but a more robust test would obtain `NUM_OF_RATINGS` from test settings and test the two so-called "edge" cases. Verify that a rating value greater than the range allows, and one less than the range allows, are both rejected with a 403.

I will point out that `NUM_OF_RATINGS` is a very poor name and actually is not accurate. To me that name indicates how many different ratings are allowed. If `NUM_OF_RATINGS` is five (5), and ratings start with zero (0), I would expect a valid rating range from 0 to 4, a total of five ratings values. However the code doesn't work that way as seen in https://github.com/pinax/pinax-ratings/blob/master/pinax/ratings/views.py#L33.

Go ahead and adjust the test to prove that a rating of `NUM_OF_RATINGS` + 1 is not accepted, and we'll consider fixing the name/code inconsistency later.

Obtain the value from test settings by following the patterns seen in https://github.com/pinax/pinax-announcements/blob/master/pinax/announcements/tests/tests.py. Find references to `settings` for an example.

File: pinax/ratings/tests/test_models.py [hunk 6]
```diff
@@ -1,14 +1,14 @@
...
-from django.test import TestCase
 
 from pinax.ratings.models import Rating
 
 from .models import Car
+from .test import TestCase
 
 
-class Tests(TestCase):
+class RatingTestCase(TestCase):
```
[the remark this answers is not in the document]
Comment (kasulani) [reply pinax/ratings/tests/test_models.py]: @grahamu I have renamed the class as recommended. I think the new name is much more clearer to me and makes a lot of sense.

File: pinax/ratings/tests/test_views.py (hunk 1 above)
Comment (kasulani) [reply pinax/ratings/tests/test_views.py]: @grahamu I have converted the comments to docstrings as per pinax coding style guidelines. I have also removed the comment that was showing the code under test.

File: pinax/ratings/tests/test_views.py (hunk 2 above)
Comment (kasulani) [reply pinax/ratings/tests/test_views.py]: @grahamu the test test_post_rating_with_invalid_content_type_id is removed and test_post_rating_with_invalid_content_type_id has been renamed

File: pinax/ratings/tests/test_views.py (hunk 5 above)
Comment (kasulani) [reply pinax/ratings/tests/test_views.py]: @grahamu I have added the two edge cases test to verify that a rating value greater than the range allows, and one less than the range allows, are both rejected with a 403. The test now obtains the NUM_OF_RATINGS from the settings

File: pinax/ratings/tests/test_views.py (hunk 3 above)
Comment (kasulani) [reply pinax/ratings/tests/test_views.py]: @grahamu docstring has been added that documents what is under test

File: pinax/ratings/tests/test_views.py (hunk 4 above)
Comment (kasulani) [reply pinax/ratings/tests/test_views.py]: @grahamu the two lines have been removed

File: pinax/ratings/tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                    "rating": 4,
+                    "category": "handling"
+                },
+            )
+            self.response_403(response)
+
+    def test_post_rating_less_than_range(self):
+        """
+        Ensure view returns 403 if the rating is less than the
+        NUM_OF_RATINGS setting added to the PINAX_RATINGS_NUM_OF_RATINGS setting
```
Comment (grahamu) [inline pinax/ratings/tests/test_views.py]: "... less than the minimum rating enforced by RatingView" seems more correct. PINAX_RATINGS_NUM_OF_RATINGS doesn't say anything about the minimum rating.

File: pinax/ratings/tests/test_views.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        in settings.py
+        """
+        with self.login(self.test_user):
+            response = self.post(
+                "pinax_ratings:rate",
+                content_type_id=ContentType.objects.get_for_model(self.forester).pk,
+                object_id=self.forester.pk,
+                data={
+                    "rating": 4,
```
Comment (grahamu) [inline pinax/ratings/tests/test_views.py]: Constants like `4` in this test often cause trouble. What happens if another developer changes the setting in runtests.py to 5? This test breaks, but not for a good reason.

Rather than use a constant, import and reference the external value. In this case I suggest adding `from django.conf import settings` up top, and

```
    "rating": settings.PINAX_RATINGS_NUM_OF_RATINGS + 1
```

in the test. The test is clearer as well, showing _why_ the failure occurs... a value 1 greater than PINAX_RATINGS_NUM_OF_RATINGS is too much.

File: pinax/ratings/tests/test_views.py (hunk 3 above)
Comment (grahamu) [reply pinax/ratings/tests/test_views.py]: looks good

File: pinax/ratings/tests/test_views.py (hunk 8 above)
Comment (kasulani) [reply pinax/ratings/tests/test_views.py]: @grahamu constant has been added

File: pinax/ratings/tests/test_views.py (hunk 7 above)
Comment (kasulani) [reply pinax/ratings/tests/test_views.py]: @grahamu comment has been edited to make it clearer
